content/blog/permutate-parsers/index.md

```diff
@@ -0,0 +1,814 @@
+
+```haskell
+data Passport = Passport
+  { birthYear :: Int,
+    issueYear :: Int,
+    expirationYear :: Int,
+    height :: String,
+    hairColor :: String,
+    eyeColor :: String,
+    passportId :: String,
+    countryId :: Maybe Int
+  }
+```
+
+We'd like to write a `entryToPassport` function with this shape:
+
+```haskell
+PassportEntry -> Maybe Passport
+```
+
+Oh boy, we're really stuck now. We have a function to validate the fields,
+but that function doesn't give us a nice way to _extract_ that data and use
+it. We really did that work for nothing...
+
+One trick we could employ it to write a function like this:
+
+```haskell
+parseField ::
+  (PassportField, String) -> Maybe (PassportField, String)
+parseField tuple =
+  if isFieldValid tuple
+    then Just tuple
+    else Nothing
+```
+
+We are still reusing the validating logic, but we end up returning
+something useful instead. We are slowly migrating our code from validating
+data to parsing data!
+
+Using our new helper we can finally implement the `entryToPassport`
+function. We'll do that in two separate steps. First we'll get all the
+values of the required fields:
+
+```haskell
+getAllRequiredFields :: PassportEntry -> Maybe [String]
+getAllRequiredFields e =
+  traverse
+    ( \field -> do
+        v <- HM.lookup field e
+        (_field, text) <- parseField (field, v)
+        return text
+    )
+    requiredFields
+```
+
+The `traverse` magic ensures that we either get all the values we're
+looking for wrapped in a `Just`, or `Nothing` if any of those fields were
+invalid. Ok, we're ready to roll now!
+
+```haskell
+entryToPassport :: PassportEntry -> Maybe Passport
+entryToPassport entry = do
+  case getAllRequiredFields entry of
+    Just [byr, iyr, eyr, hgt, hcl, ecl, pid] ->
+      Just $
+        Passport
+          { birthYear = toInt byr,
+            issueYear = toInt iyr,
+            expirationYear = toInt eyr,
+            height = hgt,
+            hairColor = hcl,
+            eyeColor = ecl,
+            passportId = pid,
+            countryId = toInt <$> HM.lookup CountryId entry
+          }
+    _ ->
+      Nothing
+```
+
+This is far from ideal: we end up having to pass `String` values around,
+which need to be parsed again into the exact types that we desire. Also we
+need to pass these values into a list and hope not to mess up the ordering
+of the fields. So it's far from perfect, but we're getting somewhere.
+
+In order to use this function in our main, we replace the last line of the
+main fuction from:
+
+```haskell
+print $ length $ filter isEntryValid entries
+```
+
+to:
+
+```haskell
+print $ length $ mapMaybe entryToPassport entries
+```
+
+Running
+[this](https://gist.github.com/Arkham/22cb983c61f5a6aaf54227c94d0c7a6b) on
+our test batch still returns `1`, which is a good sign we haven't broken
+anything. Still, I'm unhappy we had to parse our data twice, so let's look
+at something new.
```

Comment:
💭  IMO there's something fuzzy about the narrative here. Typically, I would expect an intermediate solution to motivate what comes afterwards, but if I'm following properly the drawback of this code is that we have to parse each field into the right type twice.

We solve that indirectly in the next section with permutating parsers, but could have done it differently too (by using a record with `Maybe`s for all fields instead of a list as the intermediate type, for example).

So my impression later, after finishing the post, is that I am not yet convinced that permutating parsers have solved any problem I had 😅.

Not sure what would be a better way to present this though. Maybe you could go straight to using parsers in the intermediate solution, show that the permutation part can get hairy and then show how to combine them with `Perm`?

It might be hard to make a smooth transition from the first solution to parsers, but I think that could also make the article a bit shorter. The second part felt a bit like a detour and I was already really eager to get to the parsers!

Reply:
Thank you so much, I've added a section about the fact that the thing I dislike the most is having these intermediate representation of data. They are problematic because that's there all the bugs are hiding: when converting between different data types! Also added this description of _shotgun parsing_:

> Shotgun parsing is a programming antipattern whereby parsing and input-validating code is mixed with and spread across processing code—throwing a cloud of checks at the input, and hoping, without any systematic justification, that one or another would catch all the “bad” cases.

Which I think is a better argument on why we should try parsing!